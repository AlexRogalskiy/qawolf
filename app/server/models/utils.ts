export const COLORS = [
  "#4545E5",
  "#C54BDE",
  "#56BBD6",
  "#8BC22D",
  "#E59C59",
  "#DA4E94",
  "#ABB3C2",
  "#667080",
];

export const buildColor = (
  teamColors: string[],
  colors: string[] = COLORS
): string => {
  const availableColor = colors.find((color) => !teamColors.includes(color));

  if (availableColor) return availableColor;

  return colors[teamColors.length % colors.length];
};
