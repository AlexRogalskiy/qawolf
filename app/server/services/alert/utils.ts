import { Suite, Trigger } from "../../types";

type BuildSuiteName = {
  suite: Suite;
  trigger: Trigger | null;
};

export const buildSuiteName = ({ suite, trigger }: BuildSuiteName): string => {
  return trigger?.name || suite.tag_names || "manually triggered";
};
