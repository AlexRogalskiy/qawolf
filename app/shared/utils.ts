export const buildTestCode = (
  url: string,
  domcontentloaded = false
): string => {
  let code =
    "const { context } = await launch();\nconst page = await context.newPage();\n";

  code += domcontentloaded
    ? `await page.goto('${url}', { waitUntil: "domcontentloaded" })\n`
    : `await page.goto('${url}');\n`;

  code += "// 🐺 QA Wolf will create code here";

  return code;
};

export const daysFromNow = (days = 0, existingDate?: number): string => {
  const date = existingDate ? new Date(existingDate) : new Date();
  date.setDate(date.getDate() + days);

  return date.toISOString();
};

// https://stackoverflow.com/a/1214753
export const minutesFromNow = (minutes = 0): string => {
  return new Date(Date.now() + minutes * 60000).toISOString();
};
