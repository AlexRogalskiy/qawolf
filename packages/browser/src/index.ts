export { Page as PuppeteerPage } from "puppeteer";
export * from "./browser";
export * from "./find";
export * from "./retry";
export * from "./strokes";
