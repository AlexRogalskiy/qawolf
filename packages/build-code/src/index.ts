export * from "./buildInitialCode";
export { stepToSelector } from "./buildSelector";
export * from "./buildVirtualCode";
export * from "./CodeReconciler";
