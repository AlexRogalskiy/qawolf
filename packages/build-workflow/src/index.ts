// export * from "./buildClickSteps";
// export * from "./buildScrollSteps";
// export * from "./buildSelectSteps";
// export * from "./buildSteps";
// export * from "./buildTypeSteps";
// export * from "./buildWorkflow";
