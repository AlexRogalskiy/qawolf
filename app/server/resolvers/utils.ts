/* eslint-disable @typescript-eslint/no-non-null-assertion */
import { Transaction } from "knex";

import { AuthenticationError } from "../errors";
import { Logger } from "../Logger";
import { findEnvironment } from "../models/environment";
import { findEnvironmentVariable } from "../models/environment_variable";
import { findSuite } from "../models/suite";
import { findTest } from "../models/test";
import { findTrigger } from "../models/trigger";
import { Team, Test, User } from "../types";

type EnsureEnvironmentAccess = {
  environment_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureEnvironmentVariableAccess = {
  environment_variable_id: string;
  logger: Logger;
  teams: Team[] | null;
  trx?: Transaction;
};

type EnsureSuiteAccess = {
  logger: Logger;
  suite_id: string;
  teams: Team[] | null;
};

type EnsureTeams = {
  logger: Logger;
  teams: Team[] | null;
};

type EnsureTeamAccess = {
  logger: Logger;
  team_id: string;
  teams: Team[] | null;
};

type EnsureTestAccess = {
  logger: Logger;
  teams: Team[] | null;
  test?: Test;
  test_id?: string;
};

type EnsureTriggerAccess = {
  logger: Logger;
  teams: Team[] | null;
  trigger_id: string;
  trx?: Transaction;
};

type EnsureUser = {
  logger: Logger;
  user: User | null;
};

/**
 * @summary Throws if `teams` is falsy
 * @returns `teams`
 */
export const ensureTeams = ({ logger, teams }: EnsureTeams): Team[] => {
  const log = logger.prefix("ensureTeams");

  if (!teams) {
    log.error("teams not provided");
    throw new AuthenticationError("no teams");
  }

  return teams;
};

export const ensureEnvironmentAccess = async ({
  environment_id,
  logger,
  teams,
  trx,
}: EnsureEnvironmentAccess): Promise<Team> => {
  const log = logger.prefix("ensureEnvironmentAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access environment", environment_id);

  const environment = await findEnvironment(environment_id, { logger, trx });
  const teamForEnvironment = teams!.find(
    (team) => environment.team_id === team.id
  );

  if (!teamForEnvironment) {
    log.error("teams", teamIds, "cannot access environment", environment_id);
    throw new AuthenticationError("cannot access environment");
  }

  return teamForEnvironment;
};

export const ensureEnvironmentVariableAccess = async ({
  environment_variable_id,
  logger,
  teams,
  trx,
}: EnsureEnvironmentVariableAccess): Promise<Team> => {
  const log = logger.prefix("ensureEnvironmentVariableAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug(
    "ensure teams",
    teamIds,
    "can access environment variable",
    environment_variable_id
  );

  const variable = await findEnvironmentVariable(environment_variable_id, {
    logger,
    trx,
  });
  const selectedTeam = teams!.find((team) => variable.team_id === team.id);

  if (!selectedTeam) {
    log.error(
      "teams",
      teamIds,
      "cannot access environment variable",
      environment_variable_id
    );
    throw new AuthenticationError("cannot access environment variable");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `suite.team_id`. Can also
 *   throw if `suite_id` doesn't match any suite.
 * @returns The team from `teams` that matches `suite.team_id`, if we didn't throw.
 */
export const ensureSuiteAccess = async ({
  logger,
  teams,
  suite_id,
}: EnsureSuiteAccess): Promise<void> => {
  const log = logger.prefix("ensureSuiteAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  log.debug("ensure teams", teamIds, "can access suite", suite_id);

  const suite = await findSuite(suite_id, { logger });

  if (!teamIds.includes(suite.team_id)) {
    log.error("teams", teamIds, "cannot access suite", suite_id);
    throw new Error("Teams cannot access suite");
  }
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `team_id`
 * @returns The team from `teams` that matches `team_id`, if we didn't throw.
 */
export const ensureTeamAccess = ({
  logger,
  team_id,
  teams,
}: EnsureTeamAccess): Team => {
  const log = logger.prefix("ensureTeamAccess");

  const teamIds = ensureTeams({ logger, teams }).map((team) => team.id);

  const selectedTeam = teams!.find((team) => team.id === team_id);
  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access team", team_id);
    throw new AuthenticationError("cannot access team");
  }

  log.debug("ensured teams", teamIds, "can access team", team_id);

  return selectedTeam;
};

/**
 * @summary Throws if `teams` is falsy or doesn't include `test.team_id`. Can also
 *   throw if `test_id` doesn't match any test.
 * @returns The team from `teams` that matches `test.team_id`, if we didn't throw.
 */
export const ensureTestAccess = async ({
  logger,
  teams,
  test,
  test_id,
}: EnsureTestAccess): Promise<Team> => {
  const log = logger.prefix("ensureTestAccess");

  const finalTeams = ensureTeams({ logger, teams });
  const teamIds = finalTeams.map((t) => t.id);

  log.debug("ensure teams", teamIds, "can access test", test?.id || test_id);

  const finalTest = test || (await findTest(test_id, { logger }));
  const team = finalTeams.find((t) => finalTest.team_id === t.id);

  if (!team) {
    log.error("teams", teamIds, "cannot access test", test?.id || test_id);
    throw new AuthenticationError("cannot access test");
  }

  return team;
};

export const ensureTriggerAccess = async ({
  logger,
  teams,
  trigger_id,
  trx,
}: EnsureTriggerAccess): Promise<Team> => {
  const log = logger.prefix("ensureTriggerAccess");

  const teamIds = ensureTeams({ teams, logger }).map((team) => team.id);
  log.debug("ensure teams", teamIds, "can access trigger", trigger_id);

  const trigger = await findTrigger(trigger_id, { logger, trx });
  const selectedTeam = teams!.find((team) => trigger.team_id === team.id);

  if (!selectedTeam) {
    log.error("teams", teamIds, "cannot access trigger", trigger_id);
    throw new AuthenticationError("cannot access trigger");
  }

  return selectedTeam;
};

/**
 * @summary Throws if `user` is falsy
 * @returns `user` if we didn't throw
 */
export const ensureUser = ({ logger, user }: EnsureUser): User => {
  const log = logger.prefix("ensureUser");

  if (!user) {
    log.error("user not provided");
    throw new AuthenticationError("no user");
  }

  return user;
};
