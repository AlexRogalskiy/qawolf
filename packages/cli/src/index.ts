#!/usr/bin/env node

import { buildTest } from "@qawolf/build-test";
// import { buildWorkflow } from "@qawolf/build-workflow";
import { Browser } from "@qawolf/browser";
import { CONFIG } from "@qawolf/config";
import { logger } from "@qawolf/logger";
import { Runner } from "@qawolf/runner";
import { Workflow } from "@qawolf/types";
import program from "commander";
import { outputFile, outputJson, readJson } from "fs-extra";
import { snakeCase } from "lodash";
import { resolve } from "path";
import { runTest } from "./runTest";
import { parseUrl, getUrlRoot } from "./utils";

// program
//   .command("build <eventsPath> <name>")
//   .description("build a test from events")
//   .action(async (eventsPath, name) => {
//     const sourcePath = resolve(eventsPath);
//     logger.verbose(`read events from ${sourcePath}`);
//     const events = await readJson(sourcePath);

//     const destPath = `${process.cwd()}/.qawolf`;
//     const formattedName = snakeCase(name);
//     const destWorkflowPath = `${destPath}/workflows/${formattedName}.json`;
//     const destTestPath = `${destPath}/tests/${formattedName}.test.js`;

//     logger.verbose(`build workflow -> ${destTestPath}`);
//     const workflow = buildWorkflow(events, formattedName);
//     await outputJson(destWorkflowPath, workflow, { spaces: " " });

//     logger.verbose(`build test -> ${destTestPath}`);
//     const test = buildTest(workflow);
//     await outputFile(destTestPath, test, "utf8");

//     process.exit(0);
//   });

let recordCommand = program
  .command("record <url> [name]")
  .description("record a workflow and create a test");

if (CONFIG.development) {
  recordCommand = recordCommand.option(
    "-e, --events",
    "record events (for debugging)"
  );
}

recordCommand.action(async (urlArgument, optionalName, cmd) => {
  const url = parseUrl(urlArgument);
  const name = optionalName || getUrlRoot(url);
  logger.verbose(`record url "${url.href}"`);

  const browser = await Browser.create({ record: true, url: url.href });
  await browser.waitForClose();

  const destPath = `${process.cwd()}/.qawolf`;
  const formattedName = snakeCase(name);

  if (cmd.events) {
    const destEventsPath = `${destPath}/events/${formattedName}.json`;
    logger.verbose(`create events "${name}" -> ${destEventsPath}`);
    await outputJson(destEventsPath, browser.events, { spaces: " " });
  } else {
    logger.verbose(`create test "${name}"`);
  }

  process.exit(0);
});

program
  .command("run [name]")
  .description("run a workflow")
  .action(async name => {
    const workflowPath = `${process.cwd()}/.qawolf/workflows/${snakeCase(
      name
    )}.json`;
    const workflow = (await readJson(workflowPath)) as Workflow;
    const runner = await Runner.create(workflow);

    await runner.run();
    await runner.close();

    process.exit(0);
  });

program
  .command("test [name]")
  .description("run a test")
  .action(async name => {
    const results = await runTest(name ? snakeCase(name) : null);
    const success = results.numFailedTestSuites < 1;

    process.exit(success ? 0 : 1);
  });

program.allowUnknownOption(false);

program.parse(process.argv);

if (!process.argv.slice(2).length) {
  console.log("\n");
  program.outputHelp();
}
