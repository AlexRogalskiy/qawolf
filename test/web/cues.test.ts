import { Browser, Page } from 'playwright';
import { launch } from '../../src/utils';
import { QAWolfWeb } from '../../src/web';
import { webScript } from '../../src/web/addScript';
import {
  Cue,
  buildSelectorForCues,
  toSelectorString,
} from '../../src/web/cues';
import { TEST_URL } from '../utils';

describe('browser tests', () => {
  let browser: Browser;
  let page: Page;

  beforeAll(async () => {
    browser = await launch({ headless: false, devtools: true });
    page = await browser.newPage();
    await page.addInitScript(webScript);
    await page.goto(`${TEST_URL}checkbox-inputs`);
  });

  afterAll(() => browser.close());

  describe('buildAttributeCues', () => {
    const buildAttributeCues = async (
      selector: string,
      attributes: string[],
      level: number,
      useAttributeName = false,
    ): Promise<Cue[]> => {
      return page.evaluate(
        ({ attributes, level, selector, useAttributeName }) => {
          const qawolf: QAWolfWeb = (window as any).qawolf;
          const element = document.querySelector(selector) as HTMLElement;

          return qawolf.buildAttributeCues({
            attributes,
            element,
            level,
            useAttributeName,
          });
        },
        { attributes, level, selector, useAttributeName },
      );
    };

    it('returns attribute cues', async () => {
      const cues = await buildAttributeCues('#single', ['data-qa', 'id'], 1);

      expect(cues).toEqual([
        { level: 1, type: 'attribute', value: '[data-qa="html-checkbox"]' },
        { level: 1, type: 'attribute', value: '[id="single"]' },
      ]);
    });

    it('uses attribute name as cue type if specified', async () => {
      const cues = await buildAttributeCues(
        '#single',
        ['data-qa', 'id'],
        1,
        true,
      );

      expect(cues).toEqual([
        { level: 1, type: 'data-qa', value: '[data-qa="html-checkbox"]' },
        { level: 1, type: 'id', value: '[id="single"]' },
      ]);
    });

    it('returns empty array if element has no specified attributes', async () => {
      const cues = await buildAttributeCues('#single', ['alt', 'title'], 1);

      expect(cues).toEqual([]);
    });
  });
});

describe('buildSelectorForCues', () => {
  it('builds selector from cues', () => {
    const cues = [
      { level: 0, type: 'class' as 'class', value: '.search-input' },

      {
        level: 1,
        type: 'attribute' as 'attribute',
        value: '[data-qa="search"]',
      },
      { level: 0, type: 'tag' as 'tag', value: 'input' },
      { level: 0, type: 'id' as 'id', value: '#search' },
    ];

    const selector = buildSelectorForCues(cues);

    expect(selector).toEqual([
      { name: 'css', body: '[data-qa="search"]' },
      { name: 'css', body: 'input.search-input#search' },
    ]);
  });
});

// describe('isMatch', () => {
//   it('returns false if multiple matches', async () => {
//     const browser = await launch();
//     const page = await browser.newPage();
//     await page.addInitScript(webScript);
//     await page.goto('https://google.com');

//     const isMatch = await page.evaluate(() => {
//       // fix this test
//       return false;
//     });
//     expect(isMatch).toBe(false);

//     await browser.close();
//   });
// });

describe('toSelectorString', () => {
  it('returns a pure CSS selector if possible', () => {
    const selectorString = toSelectorString([
      { name: 'css', body: '[data-qa="search"]' },
      { name: 'css', body: 'input.search-input' },
    ]);

    expect(selectorString).toBe('[data-qa="search"] input.search-input');
  });

  it('returns a single text selector', () => {
    const selectorString = toSelectorString([
      { name: 'text', body: '"Click Me!"' },
    ]);

    expect(selectorString).toBe('text="Click Me!"');
  });

  it('returns a mixed selector', () => {
    const selectorString = toSelectorString([
      { name: 'css', body: '.container' },
      { name: 'text', body: '"Submit"' },
    ]);

    expect(selectorString).toBe('css=.container >> text="Submit"');
  });
});
