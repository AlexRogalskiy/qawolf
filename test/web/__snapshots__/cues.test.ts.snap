// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`browser tests buildCues builds cues for a target 1`] = `
Array [
  Object {
    "level": 0,
    "type": "attribute",
    "value": "[data-qa=\\"html-checkbox\\"]",
  },
  Object {
    "level": 0,
    "type": "id",
    "value": "#single",
  },
  Object {
    "level": 0,
    "type": "tag",
    "value": "input",
  },
  Object {
    "level": 1,
    "type": "class",
    "value": ".container",
  },
  Object {
    "level": 1,
    "type": "tag",
    "value": "div",
  },
  Object {
    "level": 2,
    "type": "tag",
    "value": "div",
  },
  Object {
    "level": 3,
    "type": "id",
    "value": "#root",
  },
  Object {
    "level": 3,
    "type": "tag",
    "value": "div",
  },
  Object {
    "level": 4,
    "type": "tag",
    "value": "body",
  },
  Object {
    "level": 5,
    "type": "tag",
    "value": "html",
  },
]
`;
